File: app/src/main/java/com/example/vladimir/easyenglishlearn/ModelFactory.java
package com.example.vladimir.easyenglishlearn;

import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProvider;
import android.support.annotation.NonNull;

import com.example.vladimir.easyenglishlearn.category_edit.CategoryEditViewModel;
import com.example.vladimir.easyenglishlearn.word_selection.WordSelectionViewModel;

import java.util.HashMap;
import java.util.Map;

public class ModelFactory extends ViewModelProvider.NewInstanceFactory {

    private String mCategoryName;
    private ViewModel mViewModel;
    private static Map<String, ModelFactory> sFactoriesMap = new HashMap<>();


    private ModelFactory(String categoryName) {
        this.mCategoryName = categoryName;
    }

    public static ModelFactory getInstance(String categoryName) {
        if (!sFactoriesMap.containsKey(categoryName)) {
            sFactoriesMap.put(categoryName, new ModelFactory(categoryName));
        }
        return sFactoriesMap.get(categoryName);
    }

    @NonNull
    @Override
    @SuppressWarnings("unchecked")
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        if (modelClass == CategoryEditViewModel.class) {
            if (mViewModel == null || mViewModel.getClass() != CategoryEditViewModel.class) {
                mViewModel = new CategoryEditViewModel(mCategoryName);
            }
            return (T) mViewModel;
        } else if (modelClass == WordSelectionViewModel.class) {
            if (mViewModel == null || mViewModel.getClass() != WordSelectionViewModel.class) {
                mViewModel = new WordSelectionViewModel(mCategoryName);
            }
            return (T) mViewModel;
        }
        throw new IllegalArgumentException("Wrong ViewModel class");
    }
}

File: app/src/main/java/com/vladimir_x/easyenglishlearn/ModelFactory.kt
package com.vladimir_x.easyenglishlearn

import androidx.lifecycle.ViewModelProvider.NewInstanceFactory
import androidx.lifecycle.ViewModel
import com.vladimir_x.easyenglishlearn.category_edit.CategoryEditViewModel
import com.vladimir_x.easyenglishlearn.word_selection.WordSelectionViewModel
import com.vladimir_x.easyenglishlearn.ModelFactory
import java.lang.IllegalArgumentException
import java.util.HashMap

class ModelFactory private constructor(private val mCategoryName: String) : NewInstanceFactory() {
    private var mViewModel: ViewModel? = null

    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        if (modelClass == CategoryEditViewModel::class.java) {
            if (mViewModel == null || mViewModel!!.javaClass != CategoryEditViewModel::class.java) {
                mViewModel = CategoryEditViewModel(mCategoryName)
            }
            return mViewModel as T
        } else if (modelClass == WordSelectionViewModel::class.java) {
            if (mViewModel == null || mViewModel!!.javaClass != WordSelectionViewModel::class.java) {
                mViewModel = WordSelectionViewModel(mCategoryName)
            }
            return mViewModel as T
        }
        throw IllegalArgumentException("Wrong ViewModel class")
    }

    companion object {
        private val sFactoriesMap: MutableMap<String, ModelFactory> = HashMap()
        fun getInstance(categoryName: String): ModelFactory? {
            if (!sFactoriesMap.containsKey(categoryName)) {
                sFactoriesMap[categoryName] = ModelFactory(categoryName)
            }
            return sFactoriesMap[categoryName]
        }
    }
}